### PR TITLE
Introduce new `SSLContext` API & escalate deprecations.

### PR DESCRIPTION
**Pull request by tomchristie**
Branch for Version 1.0, so that we can incrementally merge ongoing changes and figure out at what point we want to cut a release.

- [ ] Add release notes.
  - Why choose HTTPX.
    - Sync + async support.
    - HTTP/1.1 + HTTP/2 support.
    - Command line client.
    - Strict typing.
    - Timeouts by default.
    - Usage as WSGI/ASGI test client.
    - Well defined transport API.
    - URL, Headers, QueryParams models.
    - Low-level API.
    - Roadmap for future improvements.
  - Document version changes.
  - Sponsorship callout. *(Related... I need to conduct an end-of-month appraisal)*
- [ ] Increment version.
- [ ] Merge changes.
  - [x] Improved SSL configuration API.
  - [ ] Error when conflicting transport options passed.
  - [ ] Reduced dependency footprint.
  - [ ] Added `NetworkOptions`.
  - [x] Deprecation for `proxies` escalated into removal.
  - [x] Deprecation for `app` escalated into removal.
  - [x] Removed undocumented `raw_url` interface.
  - [ ] Removed undocumented unnamed 4-tuple timeout.

---

Possible roadmap items...

* Response timeouts.
* Response max size.
* Added routing.
* Stricter typing for `auth`, `proxy` & `timeout`.
* Stricter typing for `default_encoding`.
* Stricter typing for `params`, `headers` & `cookies`.
* Stricter typing for `content`, `data`, `files`.
* Improved exception messaging for different network failure cases.

...

* DNS caching.
* Expose urlparse & friends.
* Add URLLib3Transport.
* Add AIOHttpTransport.
* Built-in WebSocket support.
* HTTP/3 support.

### CODE REVIEW

**Comment by karpetrosyan** [comment]
> Add URLLib3Transport.

I have updated the urllib3 transport from the [gist](https://gist.github.com/florimondmanca/d56764d78d748eb9f73165da388e546e?permalink_comment_id=5088020#gistcomment-5088020), and it seems ready to use as well as a good example of how transport can be implemented.

I've also worked a bit on AiohttpTransport, and hopefully, we'll see it soon. We could emphasize the possibility of using httpx as a Sans-IO library, capable of working with HTTP streams.

**Comment by RobertCraigie** [comment]
Would be very excited to see built-in websocket support!

**Comment by tomchristie** [comment]
I'm reviewing this and thinking we ought to be aiming for an interim 0.28 release here...

* Introduce @karpetrosyan's work on improving the `ssl_context = ...` API.
* Continue to include *deprecated* `cert`/verify` parameters, so we can get this merged and released.

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -211,9 +211,9 @@ this is where our previously generated `client.pem` comes in:
 ```
 import httpx
 
-proxies = {"all": "http://127.0.0.1:8080/"}
+ssl_context = httpx.SSLContext(verify="/path/to/client.pem"))
```
Comment (tomchristie) [inline .github/CONTRIBUTING.md:214]:
```suggestion
ssl_context = httpx.SSLContext()
ssl_context.load_verify_locations("/path/to/client.pem")
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,15 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## Unreleased
+
+* Added `httpx.SSLContext` class and `ssl_context` argument. (#3022)
+* Removed `cert` and `verify` arguments, you should use the `ssl_context=...` instead. (#3022)
```
Comment (tomchristie) [inline CHANGELOG.md:10]: * The `verify` and `cert` arguments continue to function, but are deprecated and will raise warnings. Use the `ssl_context = httpx.SSLContext()` API instead. (#3022, #3335)

Comment (tomchristie) [inline CHANGELOG.md:10]:
```suggestion
* The `verify` and `cert` arguments continue to function, but are deprecated and will raise warnings. Use the `ssl_context = httpx.SSLContext()` API instead. (#3022, #3335)
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,15 @@ All notable changes to this project will be documented in this file.
...
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## Unreleased
+
+* Added `httpx.SSLContext` class and `ssl_context` argument. (#3022)
+* The `verify` and `cert` arguments continue to function, but are deprecated and will raise warnings. Use the `ssl_context = httpx.SSLContext()` API instead. (#3022, #3335)
+* The deprecated `proxies` argument has now been removed.
+* The deprecated `app` argument has now been removed.
+* The `URL.raw` property has now been removed.
+* The `sniffio` project dependency has now been removed.
```
Comment (tomchristie) [inline CHANGELOG.md:14]:
```suggestion
```

Dropping this note, since it's internal.

**Comment by tomchristie** [comment]
Since we're now deprecating rather than removing `verify` and `cert`, I'd like to merge this into mainline now.
We can take a call later on if we'll issue any interim 0.28+ releases, or if we'll move directly to 1.0.

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,16 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## 0.28.0 (28th October, 2024)
+
+Version 0.28.0 introduces an `httpx.SSLContext()` class and `ssl_context` parameter.
```
Comment (tomchristie) [inline CHANGELOG.md:9]:
```suggestion
## Unreleased

Version <...> introduces an `httpx.SSLContext()` class and `ssl_context` parameter.
```

File: docs/advanced/ssl.md [hunk 5]
```diff
@@ -1,100 +1,199 @@
...
 
-If you'd like to use a custom CA bundle, you can use the `verify` parameter.
+```pycon
+>>> httpx.get("https://expired.badssl.com/")
+httpx.ConnectError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: certificate has expired (_ssl.c:997)
+```
 
-```python
-import httpx
+You can create and use new SSL context other than the default.
```
Comment (tomchristie) [inline docs/advanced/ssl.md:12]:
```suggestion
Verification is configured through [the SSL Context API](https://docs.python.org/3/library/ssl.html#ssl-contexts).
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,16 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
+## Unreleased
```
Comment (tomchristie) [inline CHANGELOG.md:7]:
```suggestion
## Version 0.28.0
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,16 @@ All notable changes to this project will be documented in this file.
...
+
+Version <...> introduces an `httpx.SSLContext()` class and `ssl_context` parameter.
```
Comment (tomchristie) [inline CHANGELOG.md:9]:
```suggestion
Version 0.28.0 introduces an `httpx.SSLContext()` class and `ssl_context` parameter.
```

**Pull request merged**